### PR TITLE
Adding support for tags

### PR DESCRIPTION
**Pull request by riccardoferretti**
Based on #302 

Add support for tags in the foam workspace.

Each note would be able to define tags in 2 ways:
- by having a `tags` property in YAML
  - syntax for this property will be lose, so all of the following will yield the same result:
    - `tags: tag1, tag2`
    - `tags: tag1 tag2`
    - `tags: [tag1, tag2]`
- by having `#tag` anywhere in the file

---

- [x] Support defining tags via YAML
- [x] Support defining tags via `#tag`
- [x] Add a new tag panel which lists all the tags in the workspace in alphabetical order
- [x] Clicking on a tag in the panel will trigger a search for the tag in the workspace
- [ ] (extra) Cmd+Clicking on a tag in the text will trigger a search for the tag in the workspace
- [ ] (extra) style tags in text (check interaction with Markdown Notes)

### CODE REVIEW

File: packages/foam-core/src/markdown-provider.ts [hunk 1]
```diff
@@ -8,26 +8,22 @@ import { Parent, Point } from 'unist';
...
-        column: 1,
-        offset: node.position!.end.offset! + 1,
-      };
-    }
+const tagsPlugin: ParserPlugin = {
+  onWillVisitTree: (tree, note) => {
+    note.tags = extractHashtags(note.source.text);
+  },
+  onDidFindProperties: (props, note) => {
+    const yamlTags = extractTagsFromProp(props.tags);
```
Comment (ingalless) [inline packages/foam-core/src/markdown-provider.ts:25]: We can probably skip the intermediary setting of this variable, and instead chain from the function?

File: packages/foam-core/src/utils/hashtags.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { isSome, isNumeric } from './core';
+import matchAll from 'string.prototype.matchall';
```
Comment (ingalless) [inline packages/foam-core/src/utils/hashtags.ts:2]: Woah, I didn't know you could do this!

File: packages/foam-core/test/markdown-provider.test.ts [hunk 3]
```diff
@@ -196,6 +196,54 @@ describe('wikilinks definitions', () => {
...
+    );
+    expect(noteA.tags).toEqual(new Set(['text', 'tags', 'care-about']));
+  });
+
+  it('can find tags as text in yaml', () => {
+    const noteA = createNoteFromMarkdown(
+      '/dir1/page-a.md',
+      `
+---
+tags: hello, world  this_is_good
```
Comment (ingalless) [inline packages/foam-core/test/markdown-provider.test.ts:217]: Does the double space here result in "world" and "this_is_good" separate tags? This is just curiousity on my part! Seems weird to me

**Review by ingalless — APPROVED** [review]
Good work @riccardoferretti! Made a couple comments, but they aren't anything that stop me approving 👍

File: packages/foam-core/src/utils/hashtags.ts (hunk 2 above)
Comment (riccardoferretti) [reply packages/foam-core/src/utils/hashtags.ts:2]: unfortunately atm we use node 10, which doesn't have `string.matchAll`, so I added this shim as a dependency (this is an actual dependency, the package name can be misleading though!). 
I don't love it, but updating to node 12 just for this seemed a bit overkill.
Nonetheless I see no problem in doing the update in the future, just wanted to keep it isolated and understand if there are possible side effects :)

File: packages/foam-core/test/markdown-provider.test.ts (hunk 3 above)
Comment (riccardoferretti) [reply packages/foam-core/test/markdown-provider.test.ts:217]: I actually did it on purpose :) 
I am mixing various separators in this test and wanted to validate they can be mixed (basically I don't want to be strict about the syntax of the `tags` property, people might have different conventions)

**Pull request merged**